controllers/extensionRequests.js

```diff
@@ -14,73 +14,166 @@ const { parseQueryParams } = require("../utils/queryParser");
-    }
+      let assigneeUsername = await getUsernameElseUndefined(extensionBody.assignee);
+      let assigneeId = extensionBody.assignee;
+      if (!assigneeUsername) {
+        assigneeId = await getUserIdElseUndefined(extensionBody.assignee);
+        assigneeUsername = extensionBody.assignee;
+        extensionBody.assignee = assigneeId;
+      }
 
-    if (req.userData.id !== extensionBody.assignee && !req.userData.roles?.super_user) {
-      return res.boom.forbidden("Only assigned user and super user can create an extension request for this task.");
-    }
+      if (!assigneeId) {
+        return res.boom.badRequest("User with this id or username doesn't exist.");
+      }
 
-    const { taskData: task } = await tasks.fetchTask(extensionBody.taskId);
-    if (!task) {
-      return res.boom.badRequest("Task with this id or taskid doesn't exist.");
-    }
-    if (task.assignee !== assigneeUsername) {
-      return res.boom.badRequest("This task is assigned to some different user");
-    }
-    if (task.endsOn >= extensionBody.newEndsOn) {
-      return res.boom.badRequest("The value for newEndsOn should be greater than the previous ETA");
-    }
-    if (extensionBody.oldEndsOn !== task.endsOn) {
-      extensionBody.oldEndsOn = task.endsOn;
-    }
+      if (req.userData.id !== extensionBody.assignee && !req.userData.roles?.super_user) {
+        return res.boom.forbidden("Only assigned user and super user can create an extension request for this task.");
+      }
 
-    const prevExtensionRequest = await extensionRequestsQuery.fetchExtensionRequests({
-      taskId: extensionBody.taskId,
-      assignee: extensionBody.assignee,
-    });
-    if (prevExtensionRequest.length) {
-      return res.boom.forbidden("An extension request for this task already exists.");
+      const { taskData: task } = await tasks.fetchTask(extensionBody.taskId);
+      if (!task) {
+        return res.boom.badRequest("Task with this id or taskid doesn't exist.");
+      }
+      if (task.assignee !== assigneeUsername) {
+        return res.boom.badRequest("This task is assigned to some different user");
+      }
+      if (task.endsOn >= extensionBody.newEndsOn) {
+        return res.boom.badRequest("The value for newEndsOn should be greater than the previous ETA");
+      }
+      if (extensionBody.oldEndsOn !== task.endsOn) {
+        extensionBody.oldEndsOn = task.endsOn;
+      }
+
+      const latestExtensionRequest = await extensionRequestsQuery.fetchLatestExtensionRequest({
+        taskId: extensionBody.taskId,
+      });
+
+      if (latestExtensionRequest[0] && latestExtensionRequest[0].status === "PENDING") {
+        return res.boom.badRequest("An extension request for this task already exists.");
+      }
+
+      // dont allow newETA<taskEndsOnETA
+      if (latestExtensionRequest[0]) {
+        if (latestExtensionRequest[0].status === "APPROVED" || latestExtensionRequest[0].status === "DENIED") {
```

Comment:
This if-block will always be true as status can only be [PENDING, APPROVED, DENIED] and there's already a if-block before this which returns bad request when the status is PENDING.

Reply:
Made the changes.